chain/tendermint/src/trigger.rs

```diff
@@ -0,0 +1,125 @@
+        }
+
+        let trigger_without_block = match self {
+            TendermintTrigger::Block(_) => MappingTriggerWithoutBlock::Block,
+            TendermintTrigger::Event(event) => MappingTriggerWithoutBlock::Event {
+                event_type: &event.event().event_type,
+            },
+        };
+
+        write!(f, "{:?}", trigger_without_block)
+    }
+}
+
+impl MappingTrigger for TendermintTrigger {
+    fn to_asc_ptr<H: AscHeap>(self, heap: &mut H) -> Result<AscPtr<()>, DeterministicHostError> {
+        Ok(match self {
+            TendermintTrigger::Block(event_list) => asc_new(heap, event_list.as_ref())?.erase(),
+            TendermintTrigger::Event(event_data) => asc_new(heap, event_data.as_ref())?.erase(),
+        })
+    }
+}
+
+#[derive(Clone)]
+pub enum TendermintTrigger {
+    Block(Arc<codec::EventList>),
+    Event(Arc<codec::EventData>),
+}
+
+impl CheapClone for TendermintTrigger {
+    fn cheap_clone(&self) -> TendermintTrigger {
+        match self {
+            TendermintTrigger::Block(event_list) => {
+                TendermintTrigger::Block(event_list.cheap_clone())
+            }
+            TendermintTrigger::Event(event_data) => {
+                TendermintTrigger::Event(event_data.cheap_clone())
+            }
+        }
+    }
+}
+
+impl PartialEq for TendermintTrigger {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Block(a_ptr), Self::Block(b_ptr)) => a_ptr == b_ptr,
+            (Self::Event(a), Self::Event(b)) => a.event().event_type == b.event().event_type,
+            _ => false,
+        }
+    }
+}
+
+impl Eq for TendermintTrigger {}
+
+impl TendermintTrigger {
+    pub fn block_number(&self) -> BlockNumber {
+        match self {
+            TendermintTrigger::Block(event_list) => event_list.block().number(),
+            TendermintTrigger::Event(event_data) => event_data.block().number(),
+        }
+    }
+
+    pub fn block_hash(&self) -> BlockHash {
+        match self {
+            TendermintTrigger::Block(event_list) => event_list.block().hash(),
+            TendermintTrigger::Event(event_data) => event_data.block().hash(),
+        }
+    }
+}
+
+impl Ord for TendermintTrigger {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self, other) {
+            // Keep the order when comparing two block triggers
+            (Self::Block(..), Self::Block(..)) => Ordering::Equal,
+
+            // Block triggers always come last
+            (Self::Block(..), _) => Ordering::Greater,
+            (_, Self::Block(..)) => Ordering::Less,
+
+            // Events have no intrinsic ordering information, so we keep the order in
+            // which they are included in the `events` field
+            (Self::Event(..), Self::Event(..)) => Ordering::Equal,
+        }
+    }
+}
+
+impl PartialOrd for TendermintTrigger {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
```

Comment:
it's not possible because the `TendermintTrigger` variant fields are generated by `prost` and don't implement `PartialOrd`, so they can't be compared. we need to rely on the `Ord` implementation in this case.